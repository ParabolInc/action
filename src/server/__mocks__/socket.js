const socket = {
  setAuthToken: jest.fn()
};

export default socket;
