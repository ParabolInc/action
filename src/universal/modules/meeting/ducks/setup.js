import {Map as iMap, List as iList} from 'immutable';
import emailAddresses from 'email-addresses';
import * as _ from 'lodash';
import {fetchGraphQL} from '../../../utils/fetching';


// Setup 1
export const SETUP1_ADD_EMAILS_SUCCESS = 'action/meeting/SETUP1_ADD_EMAILS_SUCCESS';
export const SETUP1_ADD_EMAILS_ERROR = 'action/meeting/SETUP1_ADD_EMAILS_ERROR';
export const SETUP1_UPDATE_INVITES_FIELD = 'action/meeting/SETUP1_UPDATE_INVITES_FIELD';

// Setup 2
export const SETUP2_REMOVE_INVITEE = 'action/meeting/SETUP2_REMOVE_INVITEE';
export const SETUP2_UPDATE_ROW_HOVER = 'action/meeting/SETUP2_UPDATE_ROW_HOVER';
export const SETUP2_INVITES_ERROR = 'action/meeting/SETUP2_INVITES_ERROR';
export const SETUP2_INVITES_SUCCESS = 'action/meeting/SETUP2_INVITES_SUCCESS';

const initialState = iMap({
  setup1: iMap({
    emails: iList(),
    invitesField: '',
    invitesFieldHasValue: false,
    invitesFieldError: '',
    invitesFieldHasError: false
  }),
  setup2: iMap({
    rowWithHover: ''
  })
});

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case SETUP1_ADD_EMAILS_ERROR:
      return state.mergeDeep({
        setup1: iMap({
          invitesFieldError: 'invalid email addresses',
          invitesFieldHasError: true
        })
      });
    case SETUP1_ADD_EMAILS_SUCCESS:
      return state.mergeDeep({
        setup1: iMap({
          emails: iList(action.payload),
          invitesField: '',
          invitesFieldHasValue: false,
          invitesFieldError: '',
          invitesFieldHasError: false
        })
      });
    case SETUP1_UPDATE_INVITES_FIELD:
      return state.mergeDeep({
        setup1: iMap({
          invitesField: action.payload.value,
          invitesFieldHasValue: action.payload.hasValue,
          invitesFieldError: '',
          invitesFieldHasError: false
        })
      });
    case SETUP2_REMOVE_INVITEE: {
      const idx = action.payload;
      const list = state.getIn(['setup1', 'emails']);
      return state.mergeDeep({
        setup1: iMap({
          emails: [
            ...list.slice(0, idx),
            ...list.slice(idx + 1)
          ]
        })
      });
    }
    case SETUP2_UPDATE_ROW_HOVER:
      return state.mergeDeep({
        setup2: iMap({
          rowWithHover: action.payload
        })
      });
    default:
      return state;
  }
}

export const addInvitesFromInvitesField = (emailsString) => {
  let emails = emailAddresses.parseAddressList(emailsString);
  if (emails === null) {
    return ({
      type: SETUP1_ADD_EMAILS_ERROR,
      payload: null
    });
  }
  emails = _.map(emails, (em) => _.pick(em, ['name', 'address']));
  return ({
    type: SETUP1_ADD_EMAILS_SUCCESS,
    payload: emails
  });
};

export const updateInvitesField = (value) => {
  let hasValue = true;
  if (value === '') {
    hasValue = false;
  }
  return ({
    type: SETUP1_UPDATE_INVITES_FIELD,
    payload: {
      value,
      hasValue
    }
  });
};

export const removeInvitee = (nameOrEmail) =>
  (dispatch, getState) => {
    const idx = getState()
      .getIn(['meetingModule', 'setup', 'setup1', 'emails'])
      .findIndex((em) => em.name === nameOrEmail || em.address === nameOrEmail);
    return dispatch({
      type: SETUP2_REMOVE_INVITEE,
      payload: idx
    });
  };

// TODO: batch this query?
export const createInvites = (teamId) =>
  async (dispatch, getState) => {
    const queries = [];
    const invitees = getState()
      .getIn(['meetingModule', 'setup', 'setup1', 'emails']);
    for (const invitee of invitees) {
      const {address} = invitee;
      const name = invitee.name || '';
      const query = `
        mutation($teamId: ID!, $name: String!, $email: String!) {
          payload: createTeamMember(teamId: $teamId, name: $name, email: $email) {
            id,
            name,
            email
          }
        }`;
      queries.push(
        fetchGraphQL({query, variables: {teamId, name, email: address}})
      );
    }
    const results = await Promise.all(queries);
    const error = _.reduce(
      results,
      (agg, result) => (agg || result.error),
      false
    );
    if (error) {
      return dispatch({ type: SETUP2_INVITES_ERROR });
    }
    return dispatch({ type: SETUP2_INVITES_SUCCESS });
  };
