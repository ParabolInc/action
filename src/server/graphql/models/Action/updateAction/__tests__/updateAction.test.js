import getRethink from 'server/database/rethinkDriver';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndTrim from 'server/__tests__/utils/fetchAndTrim';
import TrimSnapshot from 'server/__tests__/utils/TrimSnapshot';
import MockDB from 'server/__tests__/setup/MockDB';
import expectAsyncToThrow from 'server/__tests__/utils/expectAsyncToThrow';
import socket from 'server/__mocks__/socket';
import updateAction from 'server/graphql/models/Action/updateAction/updateAction';

MockDate.set(__now);
console.error = jest.fn();

describe('updateAction', () => {
  test('updates the sortOrder without changing updatedAt', async() => {
    // SETUP
    const r = getRethink();
    const trimSnapshot = new TrimSnapshot();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedAction = {
      id: actionId,
      sortOrder: 2
    };
    await updateAction.resolve(undefined, {updatedAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndTrim({
      action: r.table('Action').get(actionId),
    }, trimSnapshot);
    expect(db).toMatchSnapshot();
  });

  test.only('updates the content of the action', async() => {
    // SETUP
    const r = getRethink();
    const trimSnapshot = new TrimSnapshot();
    const mockDB = new MockDB();
    const {action, user} = await mockDB.init()
      .newAction();
    const actionId = action[0].id;
    const authToken = mockAuthToken(user[7]);

    // TEST
    const updatedAction = {
      id: actionId,
      content: 'Updated content'
    };
    await updateAction.resolve(undefined, {updatedAction}, {authToken, socket});

    // VERIFY
    const db = await fetchAndTrim({
      action: r.table('Action').get(actionId),
    }, trimSnapshot);
    expect(db).toMatchSnapshot();
  });

  test('throws when no websocket is present', async() => {
    const authToken = {};
    await expectAsyncToThrow(updateAction.resolve(undefined, {updatedAction: {}}, {authToken}));
  });
});
