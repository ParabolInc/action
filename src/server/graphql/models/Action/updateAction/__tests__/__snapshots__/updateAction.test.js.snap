// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateAction throw when the caller is not a team member 1`] = `"{\\"_error\\":\\"You do not have access to that team!\\"}"`;

exports[`updateAction throws when no websocket is present 1`] = `"{\\"_error\\":\\"this must be called from a websocket\\"}"`;

exports[`updateAction updates the content of the action 1`] = `
Object {
  "action": Object {
    "agendaId": 1,
    "content": "Updated content",
    "createdAt": 2017-01-01T07:00:00.000Z,
    "createdBy": 2,
    "id": 0,
    "sortOrder": 0,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 2,
  },
}
`;

exports[`updateAction updates the sortOrder without changing updatedAt 1`] = `
Object {
  "action": Object {
    "agendaId": 1,
    "content": "Test Action",
    "createdAt": 2017-01-01T07:00:00.000Z,
    "createdBy": 2,
    "id": 0,
    "sortOrder": 2,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T07:00:00.000Z,
    "userId": 2,
  },
}
`;

exports[`updateAction updates the status of the action 1`] = `
Object {
  "action": Object {
    "agendaId": 1,
    "content": "Test Action",
    "createdAt": 2017-01-01T07:00:00.000Z,
    "createdBy": 2,
    "id": 0,
    "isComplete": true,
    "sortOrder": 0,
    "teamMemberId": 3,
    "updatedAt": 2017-01-01T08:00:00.000Z,
    "userId": 2,
  },
}
`;
